### PR TITLE
raw_fix: Modified python function parsing to be similar to bash sourcing

### PR DESCRIPTION
**Pull request by vimalmanohar**
#569 @vijayaditya

### CODE REVIEW

File: egs/wsj/s5/steps/nnet3/nnet3_train_lib.py [hunk 1]
```diff
@@ -175,25 +175,22 @@ def SplitData(data, num_jobs):
...
+                variables['model_right_context'] = int(field_value)
             elif field_name == 'num_hidden_layers':
-                num_hidden_layers = int(field_value)
-
-        if model_left_context is not None and model_right_context is not None and num_hidden_layers is not None:
-            return [model_left_context, model_right_context, num_hidden_layers]
-
+                variables['num_hidden_layers'] = int(field_value)
+            else:
+                variables[field_name] = field_value
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/nnet3_train_lib.py]: You could handle the booleans too.. if someone did

if add_lda:

it would evaluate to true if the add_lda was either "true" or "false".
better convert all booleans to python bools.

File: egs/wsj/s5/steps/nnet3/train_rnn.py [hunk 2]
```diff
@@ -500,7 +500,15 @@ def Train(args, run_opts):
     config_dir = '{0}/configs'.format(args.dir)
     var_file = '{0}/vars'.format(config_dir)
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: could you please also add the changes to steps/nnet3/chain/train.py

Comment (vimalmanohar) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: I already made changes in that.

**Comment by vimalmanohar** [comment]
I have added add_lda option too.

File: egs/wsj/s5/steps/nnet3/nnet3_train_lib.py [hunk 3]
```diff
@@ -175,25 +183,28 @@ def SplitData(data, num_jobs):
...
-                model_right_context = int(field_value)
+                variables['model_right_context'] = int(field_value)
             elif field_name == 'num_hidden_layers':
-                num_hidden_layers = int(field_value)
-
-        if model_left_context is not None and model_right_context is not None and num_hidden_layers is not None:
-            return [model_left_context, model_right_context, num_hidden_layers]
-
+                variables['num_hidden_layers'] = int(field_value)
+            elif field_name == 'add_lda':
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/nnet3_train_lib.py]: better do a more generic conversion. If field_value is true/false convert it to python bool

File: egs/wsj/s5/steps/nnet3/nnet3_train_lib.py [hunk 4]
```diff
@@ -175,25 +183,28 @@ def SplitData(data, num_jobs):
                                                                   num_jobs = num_jobs))
 
 def ParseModelConfigVarsFile(var_file):
+    variables = {}
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/nnet3_train_lib.py]: guys- I don't see why you are adding this huge mechanism.  Didn't the old method work fine?  And didn't we only want to parse a few specific variables and ignore the rest?
Is there a specific projected use for this?

**Comment by danpovey** [comment]
I don't think this is the right direction.
I previously made a fix to this function in one place, to move a conversion to 'int' inside the if-statements (after you had checked the variable name).  It was just a change to a handful of lines.  That worked.
You are adding tons of code and I don't see that it adds any benefit.  Unless you have a good argument, I don't think you should pursue this.

**Comment by vijayaditya** [comment]
In his new raw_nnet scripts
<https://github.com/kaldi-asr/kaldi/blob/master/egs/wsj/s5/steps/nnet3/tdnn/train_raw_nnet.sh#L166>
Vimal dumps additional info like objective_type in the <model>/config/vars
file. So we started supporting additional info in this file.

Also, as we moved config generation outside the training script, it would
be good to ensure that the config/vars is consistent with the arguments
provided to the training script. e.g. we could perform checks like
include-log-softmax is true and --l2-regularization is non-zero , so mostly
there is some error in the user script.

--Vijay

**Comment by danpovey** [comment]
In his new raw_nnet scripts
> <
> https://github.com/kaldi-asr/kaldi/blob/master/egs/wsj/s5/steps/nnet3/tdnn/train_raw_nnet.sh#L166
> >
> Vimal dumps additional info like objective_type in the <model>/config/vars
> file. So we started supporting additional info in this file.
>
> Also, as we moved config generation outside the training script, it would
> be good to ensure that the config/vars is consistent with the arguments
> provided to the training script. e.g. we could perform checks like
> include-log-softmax is true and --l2-regularization is non-zero , so mostly
> there is some error in the user script.
>

I think it would be better to support a specific list of variable names,
extending the list as needed, and set any unset ones to 'None' if needed.
That way the calling code is much simpler.

Dan

**Comment by danpovey** [comment]
e.g. instead of returning [left-context, right-context, num-layers] the
function would return
[left-context, right-context, num-layers, l2-regularize].
In your proposed code you are pushing the hassle onto the outer function.

**Comment by danpovey** [comment]
Vijay, I see that this PR started because you asked Vimal to make sure his changed did not cause ParseModelConfigVars to crash.
I had already changed it to make sure it didn't crash, and it works fine.  I think I made the same changes in the chain lib.
Right now I think there is no reason to make any changes at all, unless there is a specific bug that you are trying to solve that I don't know about.

**Comment by vijayaditya** [comment]
The script in the current state does not support parsing the add_lda option
and other variables being added by the config_generator (
https://github.com/kaldi-asr/kaldi/blob/master/egs/wsj/s5/steps/nnet3/tdnn/make_configs.py#L595).
So the script crashes when used in master branch. This was identified from
Tom's attempts to run CNN experiments with xent training.

You were probably looking at the chain branch where these changes can't be
seen, as I removed them temporarily to avoid some issues being caused by
conflicts between vimal's tdnn/make_configs.py  changes and my changes.

--Vijay

**Comment by danpovey** [comment]
Can you show me how it crashes?

**Comment by danpovey** [comment]
BTW, I fixed that crash previously by moving the conversion to 'int' inside
the if statements, and that fix is in master too.  So if it was crashing I
suspect it was due to out-of-date code.

**Comment by vijayaditya** [comment]
OK. In that case this is not immediately needed. 

However we might still take advantage of the variables being dumped in the config/vars file to perform checks (include-logsoftmax and l2-regularize) and avoid stages (e.g. avoid PreconditioningMatrix computation if add_lda is false). 

So Vimal could you just modify the code to return multiple values ?
[left-context, right-context, num-layers,  add_lda, l2-regularize].

**Comment by vimalmanohar** [comment]
This is actually infeasible and not back-compatible. What do you do if the configs do not have the options add_lda and l2_regularize? Also different dnn architectures may have different of variables. I suppose add_lda is used for raw nnet and CNN, but not the other architectures. l2_regularizer does not make sense in non-chain system.

**Comment by vijayaditya** [comment]
hmm.. this discussion is just going around in circles.

As Dan suggested it is better to do the error checking in the function rather than the calling code, to keep the training scripts simple. As the chain/xent can decide which values they want they could call an appropriate parsing function, which parses the values and does the appropriate checks.

You could close this PR for now, I will just add necessary functions when we want them.

**Comment by danpovey** [comment]
My proposal was to just leave the variables as None in that case.

Dan

**Comment by vijayaditya** [comment]
I guess Vimal's concern is that we would have to change the function calls in all scripts whenever we want additional values in one script, so he was using the dictionary.

**Comment by vimalmanohar** [comment]
If the issue is that the training script looks "complicated", the conversion from dictionary to array can be done by a separate function as Vijay suggested. This solves the issue of needing to change function in all the script when we add additional values. This is a recurrence of the previously solved issue. https://github.com/kaldi-asr/kaldi/pull/522

**Comment by danpovey** [comment]
How about something like
[a,b,c] = ParseConfigFile(config_file, 'a', 'b', 'c')
and have it always just return strings, or None if not defined?  Then the
use can handle the conversion to int or bool or whatever?

Dan

**Comment by vijayaditya** [comment]
I think it would be easier to return a dictionary, the user could specify the variables they need and the function could check for the existence of these variables.

<code>value_dict = ParseConfigFile(config_file, ['a','b','c'])</code> value_dict would have 'a','b' and 'c'.

**Comment by danpovey** [comment]
If you return a dictionary then the user has to do lookup anyway so there
is no reason to pass the args.
OK, do it the way you wanted.  But call it something different.
Dan

**Comment by vijayaditya** [comment]
The args being passed to ensure that the function checks for the existence of these variables.

**Comment by vimalmanohar** [comment]
One approach would be to do something similar to reading Nnet3 component configs.

[a,b,c] = ParseVariables(ParseConfigFile(config_file)),
where the ParseConfigFile is common to all files and returns a dictionary.
ParseVariables reads specific things out of it and return an array. ParseVariables is defined in the train scripts and hence it can return the variables in their appropriate type. e.g. left_context as integer.

**Comment by vijayaditya** [comment]
@vimalmanohar lets take this discussion offline. We might be spamming the list.
